selectel/iam.go

```
@@ -162,3 +170,51 @@ func convertIAMFederationToList(federation *users.Federation) []interface{} {
 		},
 	}
 }
+
+func warnAboutDeprecatedRoles(ctx context.Context, meta interface{}, assignedRoles []roles.Role) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	if len(assignedRoles) == 0 {
+		return diags
+	}
+
+	deprecatedRoles := getDeprecatedRolesCache(ctx, meta)
+	if deprecatedRoles == nil {
+		return diags
+	}
+
+	for _, assignedRole := range assignedRoles {
+		if deprecatedRoles[assignedRole.RoleName] {
+			diags = append(diags, diag.Diagnostic{
+				Severity: diag.Warning,
+				Summary:  "Deprecated permission in use",
+				Detail:   fmt.Sprintf("Deprecated permission %q is in use. See documentation for the updated replacement.", assignedRole.RoleName),
+			})
+		}
+	}
+
+	return diags
+}
+
+func getDeprecatedRolesCache(ctx context.Context, meta interface{}) map[string]bool {
+	deprecatedRolesCacheOnce.Do(func() {
+		iamClient, diagErr := getIAMClient(meta)
+		if diagErr != nil {
```

Why are we ignoring errors here? If I understand correctly, this could lead to a situation where roles cannot be fetched from IAM for some reason, and in that case there would be no warnings when applying IAM Terraform resources – even if the customer is trying to apply a deprecated role.

Wouldn't it be better to propagate the error instead?